### DISCUSSION TITLE
Looking to indiscriminately claim chunks for a town

### DISCUSSION THREAD

**Discussion by TriageONE** [TownyAdvanced/Towny#5349] [Q&A]
I have been working on something for a little while, and i wanted to know the proper way to skip all checks and claim a chunk, whether it borders your lands or not, you have the money, you've hit your max claims, etc... 
I just want a method to claim the residential chunk as my town's own. Here is a snippet of my code, the part with the action of claiming a town block indiscriminately:

`townBlock.setTown(resident.getTown());
townBlock.setResident(resident);
townBlock.setType(TownBlockType.RESIDENTIAL);
townBlock.save();`

and i also have this, to reset the cache on the plugin

`Towny plugin = new Towny();
plugin.resetCache();`

So far, running this does not work. I know for a fact, the following: 
- resident object from event returns proper resident
- resident.getTown() returns proper town
- The claims show up in my town if i just do /t, showing like 5 of 8 instead of 1 of 8 
- the plots claimed are still wilderness. 

What am i doing wrong, and do i need more steps?

**Comment by LlmDl** [comment]
So the way Towny handles a town claiming land in the TownClaim task:
```java
			TownBlock townBlock = new TownBlock(worldCoord.getX(), worldCoord.getZ(), worldCoord.getTownyWorld());
			townBlock.setTown(town);
			// Set the plot permissions to mirror the towns.
			townBlock.setType(townBlock.getType());
			townBlock.save();
```
Covers the most basic aspect of it. Without seeing your code I cannot say what's going on though.

**Reply by TriageONE** [reply]
sorry i haven't gotten back, but i have just discovered something interesting in refrence to this. 
My implementation has a list of chunks that we will also want to claim alongside the one the player is in. 
Once the player claims that chunk, a few more around him also claim
The problem is that now, i can claim these chunks, but the server has to restart in order to acknowledge the fact that those chunks changed. 
Why would this be?

I am using your code above, and am choosing an arbitrary worldcoord that I am not standing in.

**Reply by LlmDl** [reply]
Could be very useful to share your code. The code I wrote above should be fine.

**Reply by TriageONE** [reply]
here is the current dump of code. It has a lot of useless stuff. 

    **REDACTED**

what the loop **"while (Objects.requireNonNull(childnodes).next())"** does here, is process a resultSet filled with chunk coordinates, and starting with the line
 **if (TownyAPI.getInstance().getTownBlock(worldCoord) != null)**
 We check if the location has a valid claim in session, and if it does, delete it first. 
 
 Next, we have to create a new townblock object to operate on, now that we have made it so we know that the chunk is vacant
 **TownBlock townBlock = new TownBlock(Integer.parseInt(xs.toString()), Integer.parseInt(zs.toString()), event.getTownBlock().getWorld());**
 we create a new town block with those same coordinates, and then apply the following parameters to it: 
 
  townBlock.setTown(resident.getTown()); to override ownership
  townBlock.setType(townBlock.getType()); to make type "Default"
  
  After that, i attempt to use your save function to save the operation with 
  **townBlock.save();**
  
  The worrying thing is that i have debug mode on but never see the string **TownyMessaging.sendDebugMsg("Saving town block " + townBlock.getWorld().getName() + ":" + townBlock.getX() + "x" + townBlock.getZ());**, which is part of your method ***public synchronized boolean saveTownBlock(TownBlock townBlock)*** within class ***TownySQLSource.class*** displayed in console. This is a mild worry, however it is more than likely irrelevant because of the next point. 
  
Like i said before, this algorithm works- only when the chunks that are to be claimed do not have to be overwritten over another town, whether enemy, neutral, or ally. 
IF you are in the wild, and the other chunks are too, everything works as expected
IF the chunk(s) have a current affiliation of any kind, belonging to anyone else, this claim for each of those overlapping TownBlocks fails. The only way to see an updated **/towny map**, or updated claim by walking into the chunk and watching for your splash text above your hotbar to show the correct claim, is to restart the server. 

I hope this is enough data, please let me know if you require a video showing more detail

**Reply by LlmDl** [reply]
So even if a town has already claimed the townblock, using townblock.setTown(town) will remove the previous town. You might try skipping your:
```java
                WorldCoord worldCoord = new WorldCoord("world", Integer.parseInt(xs.toString()), Integer.parseInt(zs.toString()));
                if (TownyAPI.getInstance().getTownBlock(worldCoord) != null){
                    TownBlock deleted = TownyAPI.getInstance().getTownBlock(worldCoord);
                    player.sendMessage("Attempting to remove town block");
                    townyUniverse.getDataSource().deleteTownBlock(deleted);
                    player.sendMessage("TownBlock Removed");
                    deleted.save();
                    player.sendMessage("Deletion Saved\n");
                }
```
And instead get the existing townblock, (only making a new one if the townblock doesn't exist.)
